### DISCUSSION TITLE
Static assets 404 during deployment

### DISCUSSION THREAD

**Discussion by TrevorBurnham** [vercel/next.js#66717] [Help]
### Summary

I'm building and deploying a Next.js app via a containerized process, so each deployment only includes the static assets from the last build. The problem is that, during the deployment process, there's a short window of time where the browser can load the page from the previous build, request the static JS and CSS assets for that page, and get 404s because those requests are routed to the new build.

I searched and found one previous discussion on this: https://github.com/vercel/next.js/discussions/26870 I also found this Stack Overflow question: https://stackoverflow.com/questions/68732117/in-nextjs-how-to-keep-files-from-previous-build-to-prevent-missing-file-error But I haven't come across a good solution. Copying over the static assets from the previous build is one possible solution, but it seems difficult to implement in practice.

I'm considering inserting a `<script>` tag into the page to trigger a reload if any of the page's assets fail to load. Has anyone found a better solution?

**Comment by versecafe** [comment]
Skew protection assuming you're on Vercel should solve this by keeping clients on the correct assets for up to 12 hours past a new deployment, which should cover almost every case where this would be a problem you can find it at `https://vercel.com/[team]/[project]/settings/advanced` but it only applies to pro and enterprise plans.

**Reply by versecafe** [reply]
Relevant docs: https://vercel.com/docs/deployments/skew-protection

**Reply by TrevorBurnham** [reply]
Thanks, but I’m looking for a solution that doesn’t require using Vercel’s hosting service.

**Reply by versecafe** [reply]
If you don't mind bundling two Next instances in the container Current + Prior you can have a 404 handler that then attempts to check for the file on the prior instance, this only works across one version change and has a small bit of extra latency but it's better than nothing

**Reply by versecafe** [reply]
or you can tag all requests with a git hash header and check that when hitting the middleware to redirect instead of hitting 404 first

**Reply by TrevorBurnham** [reply]
Hmm. I suppose it's possible to do something in the CI like:

1. Check out the previous commit and build those static assets.
2. Check out the current commit, build the current assets, and combine them.

The main tradeoff would be longer build times. Also, the page would still break during deployment if a build were deployed that wasn't based off of the commit immediately after the previous build (e.g. if more than one merge occurred between deploys, or if a rollback occurred).

**Reply by versecafe** [reply]
you can always bunch together a bunch of pror ones if you cache the old builds somewhere wouldn't effect build time too much

**Comment by nnarayen** [comment]
Hey! Unfortunately we're experiencing the exact same issue, and we also don't use Vercel for hosting. Were you able to find a successful workaround?

**Reply by TrevorBurnham** [reply]
Yes, as a workaround I've started uploading all static assets to a CDN and serving them from there instead of serving them directly from the Next.js server. Because the assets from all builds are present in the CDN, the browser can download all assets regardless of which build the HTML they're served corresponds to. (As a side benefit, serving the assets from a CDN is more performant.)

**Reply by nnarayen** [reply]
Awesome, thanks! That's the solution we were leaning towards too, but the main downside we're aware of is that clients can remain on older versions of the app for much longer now. Though backwards incompatible changes are rare, we'll have to build something to handle that more gracefully.

Appreciate it!
